pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -78,18 +79,48 @@ type LazyBinaryReader struct {
 	lazyLoadingGate gate.Gate
 	ctx             context.Context
 
-	readerMx      sync.RWMutex
-	reader        Reader
-	readerErr     error
-	readerInUse   sync.WaitGroup // Only increased when readerMx is held.
-	readerFactory func() (Reader, error)
+	loadedReaderC chan readerRequest
+	loadReqC      chan loadRequest
+	unloadReqC    chan unloadRequest
```

Comment:
Nit: I think that `C` (`ch`, `chan`, etc.) suffixes on channel variable/fields are redundant with the type system we have.